File: android/app/src/main/java/com/hypo/clipboard/transport/ConnectionStatusProber.kt
package com.hypo.clipboard.transport

import android.util.Log
import com.hypo.clipboard.data.settings.SettingsRepository
import com.hypo.clipboard.sync.DeviceKeyStore
import com.hypo.clipboard.transport.ws.WebSocketTransportClient
import com.hypo.clipboard.transport.ws.RelayWebSocketClient
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.flow.debounce
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.coroutines.Dispatchers
import java.time.Duration
import java.net.HttpURLConnection
import java.net.URL
import javax.inject.Inject
import javax.inject.Named
import javax.inject.Singleton

/**
 * Periodically probes connection status to paired devices and updates their online status.
 * Checks on app launch, when app comes to foreground, and every 1 minute.
 */
@Singleton
class ConnectionStatusProber @Inject constructor(
    private val transportManager: TransportManager,
    private val settingsRepository: SettingsRepository,
    private val deviceKeyStore: DeviceKeyStore,
    private val lanWebSocketClient: WebSocketTransportClient,
    private val cloudWebSocketClient: RelayWebSocketClient,
    @Named("cloud_ws_config") private val cloudConfig: com.hypo.clipboard.transport.ws.TlsWebSocketConfig
) {
    private val scope = CoroutineScope(SupervisorJob())
    private var peersObserverJob: Job? = null
    private var cloudStateObserverJob: Job? = null
    private var safetyTimerJob: Job? = null
    private var periodicCheckJob: Job? = null
    private var isProbing = false
    
    // Store dual status for each device (deviceId -> DeviceDualStatus)
    private val _deviceDualStatus = MutableStateFlow<Map<String, com.hypo.clipboard.ui.components.DeviceDualStatus>>(emptyMap())
    val deviceDualStatus: StateFlow<Map<String, com.hypo.clipboard.ui.components.DeviceDualStatus>> = _deviceDualStatus.asStateFlow()
    
    companion object {
        private const val TAG = "ConnectionStatusProber"
        // Safety timer interval for belt-and-suspenders (debugging builds) or as fallback
        // Only used if event-driven observation fails
        private val SAFETY_TIMER_INTERVAL_MS = Duration.ofMinutes(5).toMillis() // 5 minutes - safety net
        // Periodic check interval for actively querying peer status via cloud
        private val PERIODIC_CHECK_INTERVAL_MS = Duration.ofMinutes(10).toMillis() // 10 minutes
        // Debounce delay to avoid spamming probeConnections() on rapid state changes
        private val DEBOUNCE_DELAY_MS = 500L // 500ms debounce
    }
    
    /**
     * Start event-driven connection status probing.
     * Observes StateFlows for peers and cloud connection state changes,
     * with a safety timer as fallback.
     */
    fun start() {
        stop() // Stop any existing jobs
        
        // Initial probe on launch
        scope.launch {
            probeConnections()
        }
        
        // Peer discovery no longer drives LAN online status (WS connection only)
        
        // Event-driven: observe cloud connection state changes
        // StateFlow already deduplicates, so we only need debounce to avoid rapid-fire probes
        cloudStateObserverJob = scope.launch {
            cloudWebSocketClient.connectionState
                .debounce(DEBOUNCE_DELAY_MS)
                .collect { state ->
                    Log.d(TAG, "☁️ Cloud connection state changed: $state, triggering probe")
                    probeConnections()
                }
        }
        
        // Safety timer: belt-and-suspenders for debugging builds or as fallback
        // Only runs if event-driven observation fails
        safetyTimerJob = scope.launch {
            while (isActive) {
                delay(SAFETY_TIMER_INTERVAL_MS)
                if (isActive) {
                    Log.d(TAG, "⏰ Safety timer triggered probe (fallback)")
                    probeConnections()
                }
            }
        }
        
        // Periodic check: actively query peer status via cloud every 10 minutes
        periodicCheckJob = scope.launch {
            while (isActive) {
                delay(PERIODIC_CHECK_INTERVAL_MS)
                if (isActive) {
                    Log.d(TAG, "⏰ Periodic check triggered - querying cloud for connected peers")
                    probeConnections()
                }
            }
        }
        
        // Event-driven: trigger probe when LAN connections change (connect/disconnect)
        // Add delay when LAN connects to give peer time to establish cloud connection too
        transportManager.setLanConnectionChangeListener {
            Log.d(TAG, "🔌 LAN connection changed - triggering probe after 3 second delay")
            scope.launch {
                // Give peer 3 seconds to complete cloud connection (typical startup race)
                delay(3000)
                probeConnections()
            }
        }
    }
    
    /**
     * Stop event-driven probing
     */
    fun stop() {
        peersObserverJob?.cancel()
        peersObserverJob = null
        cloudStateObserverJob?.cancel()
        cloudStateObserverJob = null
        safetyTimerJob?.cancel()
        safetyTimerJob = null
        periodicCheckJob?.cancel()
        periodicCheckJob = null
    }
    
    /**
     * Probe connections immediately (called when app comes to foreground)
     */
    fun probeNow() {
        scope.launch {
            probeConnections()
        }
    }
    
    /**
     * Cleanup resources
     */
    fun cleanup() {
        stop()
        scope.cancel()
    }
    
    /**
     * Check network connectivity first, then server health
     */
    private suspend fun checkNetworkConnectivity(): Boolean {
        return withContext(Dispatchers.IO) {
            try {
                // Quick connectivity check - try to resolve a well-known DNS
                val url = URL("https://www.google.com")
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "HEAD"
                connection.connectTimeout = 2000 // 2 seconds - fast check
                connection.readTimeout = 2000
                connection.connect()
                connection.disconnect()
                true
            } catch (e: Exception) {
                Log.d(TAG, "🌐 Network connectivity check failed: ${e.message}")
                false
            }
        }
    }
    
    /**
     * Check server health via HTTP (fallback when WebSocket fails)
     */
    private suspend fun checkServerHealth(): Boolean {
        // First check if we have network connectivity
        if (!checkNetworkConnectivity()) {
            Log.d(TAG, "🌐 No network connectivity - server unreachable")
            return false
        }
        
        return withContext(Dispatchers.IO) {
            try {
                // Convert WebSocket URL to HTTP health endpoint
                val healthUrl = (cloudConfig.url ?: throw IllegalStateException("Cloud config URL cannot be null"))
                    .replaceFirst("wss://", "https://")
                    .replaceFirst("ws://", "http://")
                    .removeSuffix("/ws")
                    .let { url -> if (url.endsWith("/")) url else "$url/" }
                    .plus("health")
                val url = URL(healthUrl)
                val connection = url.openConnection() as HttpURLConnection
                connection.requestMethod = "GET"
                connection.connectTimeout = 3000 // 3 seconds - faster check
                connection.readTimeout = 3000
                connection.connect()
                
                val responseCode = connection.responseCode
                connection.disconnect()
                
                responseCode == 200
            } catch (e: Exception) {
                Log.w(TAG, "🏥 Server health check failed: ${e.message}")
                false
            }
        }
    }
    
    /**
     * Probe connection status for all paired devices.
     * Checks both LAN (via discovery + active WebSocket connections) and Cloud (via backend query).
     */
    private suspend fun probeConnections() {
        if (isProbing) {
            return
        }
        
        isProbing = true
        try {
            // Check network connectivity for device status updates
            val hasNetwork = checkNetworkConnectivity()
            if (!hasNetwork) {
                // Network is offline - mark all devices as offline
                val pairedDeviceIds = runCatching {
                    deviceKeyStore.getAllDeviceIds().toSet()
                }.getOrElse { emptySet() }
                val offlineStatuses = pairedDeviceIds.associateWith {
                    com.hypo.clipboard.ui.components.DeviceDualStatus(
                        isConnectedViaLan = false,
                        isConnectedViaCloud = false
                    )
                }
                _deviceDualStatus.value = offlineStatuses
                return
            }
            
            // Get paired device IDs from DeviceKeyStore (devices with encryption keys)
            val pairedDeviceIds = runCatching {
                deviceKeyStore.getAllDeviceIds().toList()
            }.getOrElse { emptyList() }
            
            // Query cloud for connected peers (if cloud is connected)
            // Privacy-preserving: only query status for devices we already know about
            val cloudConnectedDevices = if (cloudWebSocketClient.isConnected()) {
                try {
                    val connectedPeers = cloudWebSocketClient.queryConnectedPeers(peerIds = pairedDeviceIds)
                    val devicesWithNames = connectedPeers.map { peer ->
                        if (peer.name != null) "${peer.deviceId} (${peer.name})" else peer.deviceId
                    }
                    Log.d(TAG, "☁️ Cloud query returned ${connectedPeers.size} connected devices: $devicesWithNames")
                    connectedPeers.map { it.deviceId }.toSet()
                } catch (e: Exception) {
                    Log.w(TAG, "⚠️ Failed to query cloud for connected peers: ${e.message}", e)
                    emptySet()
                }
            } else {
                emptySet()
            }
            
            // Get active LAN connections (devices with active WebSocket connections)
            val activeLanConnections = transportManager.getActiveLanConnections()
            

            
            // Build dual status map for each paired device
            val dualStatuses = mutableMapOf<String, com.hypo.clipboard.ui.components.DeviceDualStatus>()
            
            for (deviceId in pairedDeviceIds) {
                // Check if device has a name - if not, it's an orphaned key from migration
                val deviceName = transportManager.getDeviceName(deviceId)
                if (deviceName == null) {
                    // Found an orphaned key (valid key in keystore, but no name in prefs)
                    // This happens if pairing was interrupted or during migration
                    // Heal the state by deleting the orphaned key
                    Log.w(TAG, "⚠️ Found orphaned key for device $deviceId (no name found). Deleting key to clean up state.")
                    runCatching {
                        deviceKeyStore.deleteKey(deviceId)
                        transportManager.forgetPairedDevice(deviceId)
                    }.onFailure { e ->
                        Log.e(TAG, "❌ Failed to delete orphaned key for $deviceId", e)
                    }
                    continue // Skip this device for status checks
                }
                
                val hasActiveLanConnection = activeLanConnections.contains(deviceId)
                val isConnectedViaLan = hasActiveLanConnection
                
                // Check Cloud status: device is in cloud-connected devices list
                val isConnectedViaCloud = cloudConnectedDevices.any { 
                    it.equals(deviceId, ignoreCase = true) || 
                    it == deviceId
                }
                
                dualStatuses[deviceId] = com.hypo.clipboard.ui.components.DeviceDualStatus(
                    isConnectedViaLan = isConnectedViaLan,
                    isConnectedViaCloud = isConnectedViaCloud
                )
                
                val deviceLabel = "$deviceId ($deviceName)"
                Log.d(TAG, "📊 Device $deviceLabel: LAN=${isConnectedViaLan}, Cloud=${isConnectedViaCloud}")
            }
            
            _deviceDualStatus.value = dualStatuses
            
            // Update TransportManager global connection state (mirrors macOS behavior)
            // This ensures the global UI badge updates correctly when peers disconnect
            val isCloudConnected = cloudWebSocketClient.isConnected()
            if (isCloudConnected) {
                transportManager.updateConnectionState(com.hypo.clipboard.transport.ConnectionState.ConnectedCloud)
            } else if (activeLanConnections.isNotEmpty()) {
                transportManager.updateConnectionState(com.hypo.clipboard.transport.ConnectionState.ConnectedLan)
            } else {
                transportManager.updateConnectionState(com.hypo.clipboard.transport.ConnectionState.Disconnected)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Error probing connections: ${e.message}", e)
        } finally {
            isProbing = false
        }
    }
}

File: shared/HypoCore/Sources/HypoCore/Transport/ConnectionStatusProber.swift
import Foundation
#if canImport(AppKit)
import AppKit
#endif
#if canImport(os)
import os
#endif
import Network // For NWPathMonitor

/// Event-driven connection status checker that updates device online status in real-time.
/// Uses periodic cloud presence refresh to avoid stale cloud-only statuses.
@MainActor
public final class ConnectionStatusProber {
    private let webSocketServer: LanWebSocketServer
    private weak var transportManager: TransportManager?
    private let transportProvider: TransportProvider?
    private var connectionTask: Task<Void, Never>?
    private var isProbing = false
    private let pathMonitor = NWPathMonitor() // Network monitor
    private let monitorQueue = DispatchQueue(label: "ConnectionStatusProber.NetworkMonitor")
    private var hasNetworkConnectivity = true // Initial assumption
    
    private let logger = HypoLogger(category: "ConnectionStatusProber")
    
    public init(webSocketServer: LanWebSocketServer, transportManager: TransportManager, transportProvider: TransportProvider? = nil) {
        self.webSocketServer = webSocketServer
        self.transportManager = transportManager
        self.transportProvider = transportProvider
        
        // Start network path monitor - listens for network changes and triggers immediate probe
        pathMonitor.pathUpdateHandler = { [weak self] path in
            guard let self = self else { return }
            Task { @MainActor [self] in
                let newStatus = path.status == .satisfied
                let oldStatus = self.hasNetworkConnectivity
                if oldStatus != newStatus {
                    self.logger.info("🌐", "Network connectivity changed: \(oldStatus) -> \(newStatus) (path status: \(path.status))")
                    self.hasNetworkConnectivity = newStatus
                    // Reconnect cloud WebSocket to use new IP address
                    if newStatus, let transportProvider = self.transportProvider as? DefaultTransportProvider {
                        let cloudTransport = transportProvider.getCloudTransport()
                        self.logger.info("🔄", "Reconnecting cloud WebSocket due to network change")
                        await cloudTransport.reconnect()
                    }
                    // Trigger probe after reconnect attempt to refresh cloud status
                    self.logger.info("🔍", "Triggering probe after network change")
                    await self.probeConnections()
                }
            }
        }
        // Monitor all network interfaces (WiFi, Ethernet, etc.)
        pathMonitor.start(queue: monitorQueue)
    }
    
    deinit {
        // Cancel tasks in deinit (non-isolated context)
        connectionTask?.cancel()
        pathMonitor.cancel() // Stop network monitor
    }
    
    /// Start event-driven connection status checking (no periodic polling)
    public func start() {
        stop() // Stop any existing task
        
        // Set up name lookup for cloud transport
        if let transportProvider = transportProvider as? DefaultTransportProvider {
            let cloudTransport = transportProvider.getCloudTransport()
            cloudTransport.setNameLookup { [weak transportManager] deviceId in
                transportManager?.deviceName(for: deviceId)
            }
        }
        
        // Initial probe on launch (immediate, no delay)
        Task { @MainActor in
            await probeConnections()
        }
        // Periodic cloud status check (mirrors Android 10-minute interval)
        connectionTask = Task { @MainActor [weak self] in
            guard let self else { return }
            while !Task.isCancelled {
                try? await Task.sleep(nanoseconds: 600_000_000_000) // 10 minutes
                if Task.isCancelled { break }
                await self.probeConnections()
            }
        }
    }
    
    /// Stop connection status checking
    public func stop() {
        connectionTask?.cancel()
        connectionTask = nil
    }
    
    /// Probe connections immediately (called on app activation)
    public func probeNow() {
        Task {
            await probeConnections()
        }
    }
    
    /// Probe connection status for all paired devices
    private func probeConnections() async {
        guard !isProbing else {
            return
        }
        
        isProbing = true
        defer { isProbing = false }
        
        // Probing connection status - no logging needed
        
        // Get TransportManager (weakly held)
        guard let transportManager = self.transportManager else {
            logger.info("⚠️ [ConnectionStatusProber] TransportManager is nil, cannot probe")
            return
        }
        
        let pairedDeviceIds = transportManager.pairedDevices.map { $0.id }
        var cloudConnectedDeviceIds = Set<String>()
        var cloudTransportConnected = false
        
        if let transportProvider = transportProvider as? DefaultTransportProvider {
            let cloudTransport = transportProvider.getCloudTransport()
            cloudTransportConnected = cloudTransport.isConnected()
            if cloudTransportConnected {
                do {
                    let connectedPeers = try await withTimeout(seconds: 3.0) {
                        await cloudTransport.queryConnectedPeers(peerIds: pairedDeviceIds)
                    }
                    cloudConnectedDeviceIds = Set(connectedPeers.map { $0.deviceId })
                    let preview = connectedPeers.prefix(5).map { $0.deviceId }.joined(separator: ",")
                    let suffix = connectedPeers.count > 5 ? ", …(+\(connectedPeers.count - 5))" : ""
                    logger.debug("☁️", "Cloud query returned \(connectedPeers.count) devices [\(preview)\(suffix)]")
                } catch {
                    logger.warning("☁️", "Cloud query timeout or error: \(error.localizedDescription)")
                }
            } else {
                logger.debug("☁️", "Cloud transport is NOT connected")
            }
        }
        
        if cloudTransportConnected {
            transportManager.updateConnectionState(.connectedCloud)
        } else if transportManager.hasLanConnections() {
            transportManager.updateConnectionState(.connectedLan)
        } else {
            transportManager.updateConnectionState(.disconnected)
        }
        
        transportManager.updateCloudConnectedDeviceIds(cloudConnectedDeviceIds)
    }

    private func withTimeout<T: Sendable>(
        seconds: TimeInterval,
        operation: @Sendable @escaping () async throws -> T
    ) async throws -> T {
        try await withThrowingTaskGroup(of: T.self) { group in
            group.addTask {
                try await operation()
            }
            group.addTask {
                try await Task.sleep(nanoseconds: UInt64(seconds * 1_000_000_000))
                throw CancellationError()
            }
            let result = try await group.next()!
            group.cancelAll()
            return result
        }
    }
}
